packages/bus-function-factory/lib/BusFunctionFactory.ts

```diff
@@ -0,0 +1,81 @@
+import type { IActorReply, IActorTest, IBusArgs } from '@comunica/core';
+import { Bus } from '@comunica/core';
+import type { ActorFunctionFactory, IActionFunctionFactory, IActorFunctionFactoryOutput } from './ActorFunctionFactory';
+
+/**
+ * Bus inspired by BusIndexed but specific for function factory.
+ *
+ * The implementation differs. In BusIndexed, each actor is indexed only once.
+ * Here, a single actor can be indexed multiple times (max 2).
+ */
+export class BusFunctionFactory
+  extends Bus<ActorFunctionFactory, IActionFunctionFactory, IActorTest, IActorFunctionFactoryOutput> {
+  protected readonly actorsIndex: Record<string, ActorFunctionFactory[]> = {};
+  protected readonly actorIdentifierFields: string[];
+  protected readonly actionIdentifierFields: string[];
+
+  public constructor(args: IBusArgs) {
+    super(args);
+  }
+
+  public override subscribe(actor: ActorFunctionFactory): void {
+    for (const actorId of this.getActorIdentifiers(actor)) {
+      let actors = this.actorsIndex[actorId];
+      if (!actors) {
+        actors = this.actorsIndex[actorId] = [];
+      }
+      actors.push(actor);
+      super.subscribe(actor);
+    }
+  }
+
+  public override unsubscribe(actor: ActorFunctionFactory): boolean {
+    let unsubscribed = true;
+    for (const actorId of this.getActorIdentifiers(actor)) {
+      const actors = this.actorsIndex[actorId];
+      if (actors) {
+        const i = actors.indexOf(actor);
+        if (i >= 0) {
+          actors.splice(i, 1);
+        }
+        if (actors.length === 0) {
+          delete this.actorsIndex[actorId];
+        }
+      }
+      unsubscribed = unsubscribed && super.unsubscribe(actor);
+    }
+    return unsubscribed;
+  }
+
+  public override publish(action: IActionFunctionFactory):
+  IActorReply<ActorFunctionFactory, IActionFunctionFactory, IActorTest, IActorFunctionFactoryOutput>[] {
+    const actionId = this.getActionIdentifier(action);
+    if (actionId) {
+      const actors = [ ...this.actorsIndex[actionId] || [], ...this.actorsIndex._undefined_ || [] ];
+      return actors.map((actor: ActorFunctionFactory):
+      IActorReply<ActorFunctionFactory, IActionFunctionFactory, IActorTest, IActorFunctionFactoryOutput> =>
+        ({ actor, reply: actor.test(action) }));
+    }
+    return super.publish(action);
+  }
+
+  protected getActorIdentifiers(actor: any): string[] {
```

Comment:
Since this is hot code, we need to make sure this is highly optimized.

For example, the isArray check might slow down things, as it may lead the V8 optimizing compiler to see the variable as untyped, which makes it more difficult for V8 to optimize this into a C struct.

The string conversion and concat may slow down things as well. Keeping things in pure object-form will probably be faster.

Reply:
You're right! High optimization is key here. If I think about it like that, I think that it might be better not to index the boolean, instead, let's only index the different function names. All actors returned from the index are still tested, and I don't think we can get a speed-up on the boolean check by doing it in the index itself.

Additionally, this allows the indexed bus and function factory bus to share more code.

Reply:
👍